doc/wwwoffled.man
.\" $Header: /home/amb/wwwoffle/doc/RCS/wwwoffled.man 2.14 2003/12/04 19:49:19 amb Exp $
.\"
.\"  WWWOFFLE - World Wide Web Offline Explorer - Version 2.8b.
.\"
.\"  Manual page for wwwoffled
.\"
.\"  Written by Andrew M. Bishop
.\"
.\"  This file Copyright 1997,98,99,2000,01,02,03 Andrew M. Bishop
.\"  It may be distributed under the GNU Public License, version 2, or
.\"  any higher version.  See section COPYING of the GNU Public license
.\"  for conditions under which this file may be redistributed.
.\"
.TH wwwoffled 8 "December 4th, 2003"

.SH NAME

wwwoffled \- A proxy server for the World Wide Web Offline Explorer.

.SH SYNOPSIS

.B wwwoffled
.B [\-h|\-\-help]
.B [\-\-version]
.B [\-c <config-file>]
.B [\-d [<log-level>]]
.B [\-l <log-file>]
.B [\-f]
.B [\-p]

.SH DESCRIPTION

.I wwwoffled
is a proxy HTTP server for the
.I
World Wide Web Offline Explorer program.  Using a standard web browser with the
HTTP proxy set to the wwwoffled server, web pages can be requested while not
connected to the internet.
.LP
When the computer is connected to the internet, the server will fetch the web
pages requested by the browser and also store them in the cache.  When the
computer is not connected, browsing of the pages in the cache is still possible,
and links can be followed.  This causes the request for the page to be stored by
the proxy server until a later time when the computer is connected and the pages
are fetched non-interactively.
.LP
Pages that are requested from a server on the same host (localhost) are not
cached, and are always fetched fresh from the server.
.LP
There is a welcome page at the URL http://localhost:8080/ that contains some
information about the program and links to the cache index, interactive refresh
page, interactive control page and WWWOFFLE internet home page.
.LP
The pages that are stored in the cache can be indexed by using the URL
http://localhost:8080/index/ to get a list of the hosts that have cached pages.
The index allows sorting of the pages into time or alphabetical order, by
following links from this page.  There is also indexes available showing the
pages that have beem modified in the last week, those fetched last time online
and those requests that are waiting to be fetched next time online.
.LP
The functions available from the wwwoffle(1) program for fetching URLs are also
available in the interactive refresh page at http://localhost:8080/refresh/.
.LP
The functions available from the wwwoffle(1) program for controlling the
wwwoffled program are also available in the interactive control page at
http://localhost:8080/control/.

.SH OPTIONS

.TP
.B \-h | \-\-help
A help message is printed giving a brief description of the usage of the
program.
.TP
.B \-\-version
The version number of the program is printed.
.TP
.B \-c <config-file>
The configuration for the program is stored in a configuration file
wwwoffle.conf(5).  This argument specifies to the program the name of that file.
Sending a HUP signal to the daemon will cause this config file to be re-read.
.TP
.B \-d [<log-level>]
Do not detach from the terminal when starting and report error messages on
standard error.  The log-level is a number from 0 (for no output on stderr) to 6
(for full debugging output on stderr).  If this is not specified then the
log-level in the config file is used.
.TP
.B \-l <log-file>
Detach from the terminal but write to a log file what would have been written to
stderr if the \fB-d\fR option had been used.  Defaults to using a log level of
4, but using the \fB-d\fR option can change this (but it will still detach).
Sending a HUP signal to the daemon will cause this log file to be closed and
re-opened.
.TP
.B \-f
Start the daemon in debugging mode (implies \fB-d\fR and overrides \fB-l\fR) and
when the first HTTP request comes in handle it without creating a child process
and then exit.
.TP
.B \-p
Print the pid of the daemon process on stdout. This option is ignored if the
\fB-d\fR or \fB-f\fR options are used.

.SH FILES

/etc/wwwoffle/wwwoffle.conf The wwwoffle.conf(5) configuration file.

.SH SEE ALSO

wwwoffle(1), wwwoffle.conf(5)

.SH AUTHOR

Andrew M. Bishop 1996,97,98,99,2000,01,02,03 (amb@gedanken.demon.co.uk)


doc/NEWS
Version 2.7f of WWWOFFLE released : Sat Sep 28 20:45:00 2002
------------------------------------------------------------

Bug Fixes:
 Fix the Set-Cookie problem again.  Convert German characters with umlauts when
 generating HTML from text.  Fix for partial decompression of gzipped pages.
 Close the client socket if the remote end cannot be determined.  Fixed namazu
 indexing script path error.  Don't attempt to modify HTML if the page is
 compressed.  Bug fix for allowing URL-SPECs to contain '='.

New Features:
 Updated German translations (most were in v2.7e also) including README.CONF.
 Added a new question to the FAQ about startup error/warning messages.
 Several new items have been added to the contrib directory.


*NOTE* See the important notes for version 2.7 listed below.


Version 2.7e of WWWOFFLE released : Sat Aug 31 11:30:00 2002
------------------------------------------------------------

Bug Fixes:
 Correct some invalid HTML in the message files.  Fix some memory leaks, free
 some unfreed memory, check some pointers, close some files.  Don't send the
 extra arguments to POST requests that WWWOFFLE uses internally.  Removed
 segmentation fault potential when password not used in config file.  Another
 fix for the cookie problem.  Warn if running as root.  Try both IPv6 and IPv4
 socket binding (IPv6 may not accept IPv4).  Potential bug fix for page
 corruption.  Potential fix for IPv6 configuration on Solaris.

New Features:
 Add an option to have case-insensitive matching for URL-SPEC path and args.
 Added the option to only fetch images on the same host (automatic fetching).
 Allow URL-SPECs to contain an '=' sign embedded in them (long time bug fix).
 The monitor options page now accepts ranges of hours or days (e.g. '1-5').


*NOTE* See the important notes for version 2.7 listed below.


Version 2.7d of WWWOFFLE released : Sun Jul 28 16:30:00 2002
------------------------------------------------------------

Bug Fixes:
 Discard POST/PUT requests that have negative content-lengths.  Make the
 CanonicaliseHost() function robust to bad IP addresses.  Fix some memory leaks,
 free some unfreed memory.  Don't give socket error using '-f' option.  Fix
 compilation on Cygwin.  Fix IPv6 compilation on Solaris 9.  Bug fix for v2.7c
 Cookie change.  Don't replace '//' in a URL path with '/'.


*NOTE* This version contains some security related fixes and should be used in
       preference to earlier versions.

*NOTE* See the important notes for version 2.7 listed below.


Version 2.7c of WWWOFFLE released : Sun Jul 7 11:00:00 2002
-----------------------------------------------------------

Bug Fixes:
 Install two DLLs for the Win32 version.  Don't crash for HTTP servers that send
 headers prefixed with whitespace.  Make the "edit selected entry" option work.
 Don't write uncompressed data to the cache with a header saying it is
 compressed.  Be more lenient in detecting spiders that cannot make requests.
 The wwwoffle-tools programs now handle dir names as if they had http:// in
 front.  Disallow wwwoffle requests for protocols that WWWOFFLE does not
 handle.  Use the command line config filename in error messages.  Fix to allow
 compilation on SGI IRIX.  Handle XHTML style tags when modifying HTML.  Updated
 setuid/setgid code.  Some memory leaks removed and potential crashes removed
 due to using lint).

New Features:
 Split up Set-Cookie headers since browsers can't handle them.  
 Don't request deflated data since WWWOFFLE and servers don't agree on format.
 Added a form on the monitor options page to stop monitoring a URL.
 The confirm-requests option now asks for confirmation for page reloads.

Documentation:
 Update FAQ to reference privoxy as well as JunkBuster.
 Describe how to modify htdig templates to work with WWWOFFLE.


*NOTE* See the important notes for version 2.7 listed below.


Version 2.7b of WWWOFFLE released : Sun Apr 21 17:45:00 2002
------------------------------------------------------------

Bug Fixes:
 Ensure that only one argument is given to wwwoffle -o or -O.  Some more version
 2.7 documentation updates.  Fix crash using 'wwwoffle -O|-o|-put|-post <URL>'.

src/wwwoffled.c
/***************************************
  $Header: /home/amb/wwwoffle/src/RCS/wwwoffled.c 2.56 2002/08/11 09:38:19 amb Exp $

  WWWOFFLE - World Wide Web Offline Explorer - Version 2.7e.
  A demon program to maintain the database and spawn the servers.
  ******************/ /******************
  Written by Andrew M. Bishop

  This file Copyright 1996,97,98,99,2000,01,02 Andrew M. Bishop
  It may be distributed under the GNU Public License, version 2, or
  any higher version.  See section COPYING of the GNU Public license
  for conditions under which this file may be redistributed.
  ***************************************/


#include "autoconfig.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <sys/types.h>
#include <unistd.h>
/* The following declarations should be contained in unistd.h, but for some reason there are not on my system.
   I list them here to avoid distracting warning messages when compiling with the -Wall option. */
#if HAVE_SETRESUID
int setresuid(uid_t ruid, uid_t euid, uid_t suid);
#endif
#if HAVE_SETRESGID
int setresgid(gid_t rgid, gid_t egid, gid_t sgid);
#endif

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#include <sys/wait.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <grp.h>

#include "version.h"
#include "wwwoffle.h"
#include "misc.h"
#include "config.h"
#include "sockets.h"
#include "errors.h"


static void usage(int verbose);
static void demoninit(void);
static void install_sighandlers(void);
static void sigchild(int signum);
static void sigexit(int signum);
static void sighup(int signum);


/*+ The server sockets that we listen on +*/
int http_fd[2]={-1,-1},          /*+ for the HTTP connections. +*/
    wwwoffle_fd[2]={-1,-1};      /*+ for the WWWOFFLE connections. +*/

/*+ The online / offline /autodial status. +*/
int online=0;

/*+ The current number of active servers +*/
int n_servers=0,                /*+ in total. +*/
    n_fetch_servers=0;          /*+ fetching a page. +*/

/*+ The maximum number of servers +*/
int max_servers=0,              /*+ in total. +*/
    max_fetch_servers=0;        /*+ fetching a page. +*/

/*+ The wwwoffle client file descriptor when fetching. +*/
int fetch_fd=-1;

/*+ The pids of the servers. +*/
int server_pids[MAX_SERVERS];

/*+ The pids of the servers that are fetching. +*/
int fetch_pids[MAX_FETCH_SERVERS];

/*+ The current purge status. +*/
int purging=0;

/*+ The pid of the purge process. +*/
int purge_pid=0;

/*+ The current status, fetching or not. +*/
int fetching=0;

/*+ The file descriptor of the sppol directory. +*/
int fSpoolDir=-1;

/*+ Set to 1 when the demon is to shutdown. +*/
int closedown=0;

/*+ Set to 1 when the demon is to re-read config file due to a sighup. +*/
static int readconfig=0;

/*+ Remember that we got a SIGCHLD +*/
static sig_atomic_t got_sigchld=0;

/*+ True if run as a demon +*/
static int detached=1;

/*+ True if the pid of the daemon should be printed at startup +*/
static int print_pid=0;

/*+ True if the -f option was passed on the command line. +*/
int nofork=0;

/* some code to save the hostname and ip address of the client in global variables */
/* static char saved_hostname[max_hostname_len],saved_ip[ipaddr_strlen]; */
extern char *client_hostname, *client_ip;
#define save_client_hostname_ip(host,ip) {client_hostname=host; client_ip=ip;}

/*++++++++++++++++++++++++++++++++++++++
  The main program.
  ++++++++++++++++++++++++++++++++++++++*/

int main(int argc, char** argv)
{
 int i;
 int err;
 struct stat buf;
 char *config_file=NULL;

 /* Parse the command line options */

 for(i=1;i<argc;i++)
   {
    if(!strcmp(argv[i],"-h") || !strcmp(argv[i],"--help"))
       usage(1);

    if(!strcmp(argv[i],"--version"))
       usage(2);

    if(!strcmp(argv[i],"-d"))
      {
       detached=0;
       if(i<(argc-1) && isdigit(argv[i+1][0]) && !argv[i+1][1])
         {
          DebuggingLevel=Fatal+1-atoi(argv[++i]);
          if(DebuggingLevel<0)
             DebuggingLevel=0;
         }
       continue;
      }

    if(!strcmp(argv[i],"-p"))
      {
       print_pid=1;
       continue;
      }

    if(!strcmp(argv[i],"-f"))
      {
       nofork=1;
       detached=0;
       continue;
      }

    if(!strcmp(argv[i],"-c"))
      {
       if(++i>=argc)
         {fprintf(stderr,"wwwoffled: The '-c' option requires a filename.\n"); exit(1);}

       config_file=argv[i];
       continue;
      }

    fprintf(stderr,"wwwoffled: Unknown option '%s'.\n",argv[i]); exit(1);
   }

 /* Initialise things. */

 for(i=0;i<MAX_FETCH_SERVERS;i++)
    fetch_pids[i]=0;

 for(i=0;i<MAX_SERVERS;i++)
    server_pids[i]=0;

 InitErrorHandler("wwwoffled",0,1); /* use stderr and not syslog to start with. */

 /* Read the configuration file. */

 InitConfigurationFile(config_file);

 init_buffer(2);
 if(ReadConfigurationFile(2))
    PrintMessage(Fatal,"Error in configuration file '%s'.",ConfigurationFileName());

 InitErrorHandler("wwwoffled",ConfigInteger(UseSyslog),1); /* enable syslog if requested. */

 /* Print a startup message. */

#if USE_ZLIB
#define ZLIB_STRING "with zlib"
#else
#define ZLIB_STRING "without zlib"
#endif

#if USE_IPV6
#define IPV6_STRING "with ipv6"
#else
#define IPV6_STRING "without ipv6"
#endif

 PrintMessage(Important,"WWWOFFLE Demon Version %s (%s,%s) started.",WWWOFFLE_VERSION,ZLIB_STRING,IPV6_STRING);
 PrintMessage(Inform,"WWWOFFLE Read Configuration File '%s'.",ConfigurationFileName());

 /* Change the user and group. */
 {
   uid_t uid=ConfigInteger(WWWOFFLE_Uid);
   gid_t gid=ConfigInteger(WWWOFFLE_Gid);

   /* gain superuser privileges if possible */
   if(geteuid()!=0 && uid!=-1) seteuid(0);

   if(gid != -1)
     {
#if HAVE_SETGROUPS
       if(geteuid()==0 || getuid()==0)
	 if(setgroups(0, NULL)<0)
	   PrintMessage(Fatal,"Cannot clear supplementary group list [%!s].");
#endif

#if HAVE_SETRESGID
       if(setresgid(gid,gid,gid)<0)
	 PrintMessage(Fatal,"Cannot set real/effective/saved group id to %d [%!s].",gid);
#else
       if(geteuid()==0)
	 {
	   if(setgid(gid)<0)
	     PrintMessage(Fatal,"Cannot set group id to %d [%!s].",gid);
	 }
       else
	 {
#if HAVE_SETREGID
	   if(setregid(getegid(),gid)<0)
	     PrintMessage(Fatal,"Cannot set effective group id to %d [%!s].",gid);
	   if(setregid(gid,-1)<0)
	     PrintMessage(Fatal,"Cannot set real group id to %d [%!s].",gid);
#else
	   PrintMessage(Fatal,"Must be root to totally change group id.");
#endif
	 }
#endif
     }

   if(uid!=-1)
     {
#if HAVE_SETRESUID
       if(setresuid(uid,uid,uid)<0)
	 PrintMessage(Fatal,"Cannot set real/effective/saved user id to %d [%!s].",uid);
#else
       if(geteuid()==0)
	 {
	   if(setuid(uid)<0)
	     PrintMessage(Fatal,"Cannot set user id to %d [%!s].",uid);
	 }
       else
	 {
#if HAVE_SETREUID
	   if(setreuid(geteuid(),uid)<0)
	     PrintMessage(Fatal,"Cannot set effective user id to %d [%!s].",uid);
	   if(setreuid(uid,-1)<0)
	     PrintMessage(Fatal,"Cannot set real user id to %d [%!s].",uid);
#else
	   PrintMessage(Fatal,"Must be root to totally change user id.");
#endif
	 }
#endif
     }

   if(uid!=-1 || gid!=-1)
     PrintMessage(Inform,"Running with uid=%d, gid=%d.",geteuid(),getegid());
 }

 if(geteuid()==0 || getegid()==0)
    PrintMessage(Warning,"Running with root user or group privileges is not recommended.");

 /* Bind the HTTP proxy socket(s). */

#if USE_IPV6
 if(ConfigString(Bind_IPv6))
   {
    http_fd[1]=OpenServerSocket(ConfigString(Bind_IPv6),ConfigInteger(HTTP_Port));
    if(http_fd[1]==-1)
       PrintMessage(Fatal,"Cannot create HTTP IPv6 server socket.");
   }
#endif

 if(ConfigString(Bind_IPv4))
   {
    http_fd[0]=OpenServerSocket(ConfigString(Bind_IPv4),ConfigInteger(HTTP_Port));
    if(http_fd[0]==-1)
      {
#if USE_IPV6
       if(http_fd[1]!=-1 && ConfigString(Bind_IPv4) && !strcmp(ConfigString(Bind_IPv4),"0.0.0.0") &&
                            ConfigString(Bind_IPv6) && !strcmp(ConfigString(Bind_IPv6),"0:0:0:0:0:0:0:0"))
          PrintMessage(Warning,"Cannot create HTTP IPv4 server socket (but the IPv6 one might accept IPv4 connections).");
       else
          PrintMessage(Fatal,"Cannot create HTTP IPv4 server socket.");
#else
       PrintMessage(Fatal,"Cannot create HTTP server socket.");
#endif
      }
   }

 if(http_fd[0]==-1 && http_fd[1]==-1)
   {
#if USE_IPV6
    PrintMessage(Fatal,"The IPv4 and IPv6 HTTP sockets were not bound; are they disabled in the config file?");
#else
    PrintMessage(Fatal,"The HTTP socket was not bound; are they disabled in the config file?");
#endif
   }

 /* Bind the WWWOFFLE control socket(s). */

#if USE_IPV6
 if(ConfigString(Bind_IPv6))
   {
    wwwoffle_fd[1]=OpenServerSocket(ConfigString(Bind_IPv6),ConfigInteger(WWWOFFLE_Port));
    if(wwwoffle_fd[1]==-1)
       PrintMessage(Fatal,"Cannot create WWWOFFLE IPv6 server socket.");
   }
#endif

 if(ConfigString(Bind_IPv4))
   {
    wwwoffle_fd[0]=OpenServerSocket(ConfigString(Bind_IPv4),ConfigInteger(WWWOFFLE_Port));
    if(wwwoffle_fd[0]==-1)
      {
#if USE_IPV6
       if(wwwoffle_fd[1]!=-1 && ConfigString(Bind_IPv4) && !strcmp(ConfigString(Bind_IPv4),"0.0.0.0") &&
                                ConfigString(Bind_IPv6) && !strcmp(ConfigString(Bind_IPv6),"0:0:0:0:0:0:0:0"))
          PrintMessage(Warning,"Cannot create WWWOFFLE IPv4 server socket (but the IPv6 one might accept IPv4 connections).");
       else
          PrintMessage(Fatal,"Cannot create WWWOFFLE IPv4 server socket.");
#else
       PrintMessage(Fatal,"Cannot create WWWOFFLE server socket.");
#endif
      }
   }

 if(wwwoffle_fd[0]==-1 && wwwoffle_fd[1]==-1)
   {
#if USE_IPV6
    PrintMessage(Fatal,"The IPv4 and IPv6 WWWOFFLE sockets were not bound; are they disabled in the config file?");
#else
    PrintMessage(Fatal,"The WWWOFFLE socket was not bound; are they disabled in the config file?");
#endif
   }

 /* Create, Change to and open the spool directory. */

 umask(0);

 if(stat(ConfigString(SpoolDir),&buf))
   {
    err=mkdir(ConfigString(SpoolDir),(mode_t)ConfigInteger(DirPerm));
    if(err==-1 && errno!=EEXIST)
       PrintMessage(Fatal,"Cannot create spool directory %s [%!s].",ConfigString(SpoolDir));
    stat(ConfigString(SpoolDir),&buf);
   }

 if(!S_ISDIR(buf.st_mode))
    PrintMessage(Fatal,"The spool directory %s is not a directory.",SpoolDir);

 err=chdir(ConfigString(SpoolDir));
 if(err==-1)
    PrintMessage(Fatal,"Cannot change to spool directory %s [%!s].",ConfigString(SpoolDir));

#if !defined(__CYGWIN__)
 fSpoolDir=open(".",O_RDONLY);
 if(fSpoolDir==-1)
    PrintMessage(Fatal,"Cannot open the spool directory '%s' [%!s].",ConfigString(SpoolDir));
#endif

 /* Detach from terminal */

 if(detached)
   {
    demoninit();
    PrintMessage(Important,"Detached from terminal and changed pid to %d.",getpid());
    InitErrorHandler("wwwoffled",1,0); /* pid changes after detaching, enable syslog, disable stderr. */
   }

 install_sighandlers();

 max_servers=ConfigInteger(MaxServers);
 max_fetch_servers=ConfigInteger(MaxFetchServers);

 /* Loop around waiting for connections. */

 do
   {
    struct timeval tv;
    fd_set readfd;
    int nfds=0,nfd=0;

    FD_ZERO(&readfd);

#if USE_IPV6
       for(nfd=0;nfd<=1;nfd++)
         {
#endif
          if(http_fd[nfd]>=nfds)
             nfds=http_fd[nfd]+1;
          if(wwwoffle_fd[nfd]>=nfds)
             nfds=wwwoffle_fd[nfd]+1;

          if(n_servers<max_servers)
             if(http_fd[nfd]!=-1)
                FD_SET(http_fd[nfd],&readfd);

          if(wwwoffle_fd[nfd]!=-1)
             FD_SET(wwwoffle_fd[nfd],&readfd);
#if USE_IPV6
         }
#endif

    tv.tv_sec=10;
    tv.tv_usec=0;

    if(select(nfds,&readfd,NULL,NULL,&tv)!=-1)
      {
#if USE_IPV6
       for(nfd=0;nfd<=1;nfd++)
         {
#endif
          if(wwwoffle_fd[nfd]!=-1 && FD_ISSET(wwwoffle_fd[nfd],&readfd))
            {
             char *host,*ip;
             int port,client;

             client=AcceptConnect(wwwoffle_fd[nfd]);
             init_buffer(client);

             if(client>=0 && !SocketRemoteName(client,&host,&ip,&port))
               {
                char *canonical_ip=CanonicaliseHost(ip);

                if((host && IsAllowedConnectHost(host)) || IsAllowedConnectHost(canonical_ip))
                  {
                   PrintMessage(Important,"WWWOFFLE Connection from host %s (%s).",host?host:"unknown",canonical_ip); /* Used in audit-usage.pl */

                   CommandConnect(client);

                   if(fetch_fd!=client)
                      CloseSocket(client);
                  }
                else
                  {
                   PrintMessage(Warning,"WWWOFFLE Connection rejected from host %s (%s).",host?host:"unknown",canonical_ip); /* Used in audit-usage.pl */
                   CloseSocket(client);
                  }

                if(canonical_ip!=ip)
                   free(canonical_ip);
               }
            }

          if(http_fd[nfd]!=-1 && FD_ISSET(http_fd[nfd],&readfd))
            {
             char *host,*ip;
             int port,client;

             client=AcceptConnect(http_fd[nfd]);
             init_buffer(client);

             if(client>=0 && !SocketRemoteName(client,&host,&ip,&port))
               {
                char *canonical_ip=CanonicaliseHost(ip);

                if((host && IsAllowedConnectHost(host)) || IsAllowedConnectHost(canonical_ip))
                  {
                   PrintMessage(Inform,"HTTP Proxy connection from host %s (%s).",host?host:"unknown",canonical_ip); /* Used in audit-usage.pl */
		   /* save hostname and ip address of client in case we need it again */
		   save_client_hostname_ip(host,ip);

                   ForkServer(client,1);
                  }
                else
                   PrintMessage(Warning,"HTTP Proxy connection rejected from host %s (%s).",host?host:"unknown",canonical_ip); /* Used in audit-usage.pl */

                if(nofork)
                   closedown=1;

                if(!nofork)
                   CloseSocket(client);

                if(canonical_ip!=ip)
                   free(canonical_ip);
               }
            }
#if USE_IPV6
         }
#endif
      }

    if(readconfig)
      {
       readconfig=0;

       PrintMessage(Important,"SIGHUP signalled.");

       PrintMessage(Important,"WWWOFFLE Re-reading Configuration File.");

       if(ReadConfigurationFile(-1))
          PrintMessage(Warning,"Error in configuration file; keeping old values.");

       PrintMessage(Important,"WWWOFFLE Finished Re-reading Configuration File.");
      }

    if(got_sigchld)
      {
       int pid, status;
       int isserver=0;

       /* To avoid race conditions, reset the flag before fetching the status */
       got_sigchld=0;

       while((pid=waitpid(-1,&status,WNOHANG))>0)
         {
          int i;
          int exitval=0;

          if(WIFEXITED(status))
             exitval=WEXITSTATUS(status);
          else if(WIFSIGNALED(status))
             exitval=-WTERMSIG(status);

          if(purging)
             if(purge_pid==pid)
               {
                if(exitval>=0)
                   PrintMessage(Inform,"Purge process exited with status %d (pid=%d).",exitval,pid);
                else
                   PrintMessage(Important,"Purge process terminated by signal %d (pid=%d).",-exitval,pid);

                purging=0;
               }

          for(i=0;i<max_servers;i++)
             if(server_pids[i]==pid)
               {
                n_servers--;
                server_pids[i]=0;
                isserver=1;

                if(exitval>=0)
                   PrintMessage(Inform,"Child wwwoffles exited with status %d (pid=%d).",exitval,pid);
                else
                   PrintMessage(Important,"Child wwwoffles terminated by signal %d (pid=%d).",-exitval,pid);

                break;
               }

          /* Check if the child that terminated is one of the fetching wwwoffles */
          for(i=0;i<max_fetch_servers;i++)
             if(fetch_pids[i]==pid)
               {
                n_fetch_servers--;
                fetch_pids[i]=0;
                break;
               }

          if(exitval==3)
             fetching=0;

          if(exitval==4 && online==1)
             fetching=1;

          if(online!=1)
             fetching=0;
         }

       if(isserver)
          PrintMessage(Debug,"Currently running: %d servers total, %d fetchers.",n_servers,n_fetch_servers);
      }

    /* The select timed out or we got a signal. If we are currently fetching,
       start fetch servers to look for jobs in the spool directory. */

    while(fetching && n_fetch_servers<max_fetch_servers && n_servers<max_servers)
       ForkServer(fetch_fd,0);

    if(fetch_fd!=-1 && !fetching && n_fetch_servers==0)
      {
       write_string(fetch_fd,"WWWOFFLE No more to fetch.\n");
       CloseSocket(fetch_fd);
       fetch_fd=-1;

       PrintMessage(Important,"WWWOFFLE Fetch finished.");

       ForkRunModeScript(ConfigString(RunFetch),"fetch","stop",fetch_fd);
      }
   }
 while(!closedown);

 /* Close down and exit. */

 if(http_fd[0]!=-1) CloseSocket(http_fd[0]);
 if(wwwoffle_fd[0]!=-1) CloseSocket(wwwoffle_fd[0]);

#if USE_IPV6
 if(http_fd[1]!=-1) CloseSocket(http_fd[1]);
 if(wwwoffle_fd[1]!=-1) CloseSocket(wwwoffle_fd[1]);
#endif

 if(!nofork)
   {
    if(n_servers)
       PrintMessage(Important,"Exit signalled - waiting for %d child wwwoffles servers.",n_servers);
    else
       PrintMessage(Important,"Exit signalled.");
   }

 while(n_servers)
   {
    int i;
    int pid,status,exitval=0;

    while((pid=waitpid(-1,&status,0))>0)
      {
       if(WIFEXITED(status))
          exitval=WEXITSTATUS(status);
       else if(WIFSIGNALED(status))
          exitval=-WTERMSIG(status);

       for(i=0;i<max_servers;i++)
          if(server_pids[i]==pid)
            {
             n_servers--;
             server_pids[i]=0;

             if(exitval>=0)
                PrintMessage(Inform,"Child wwwoffles exited with status %d (pid=%d).",exitval,pid);
             else
                PrintMessage(Important,"Child wwwoffles terminated by signal %d (pid=%d).",-exitval,pid);

             break;
            }

       if(purging)
          if(purge_pid==pid)
            {
             if(exitval>=0)
                PrintMessage(Inform,"Purge process exited with status %d (pid=%d).",exitval,pid);
             else
                PrintMessage(Important,"Purge process terminated by signal %d (pid=%d).",-exitval,pid);

             purging=0;
            }
      }
   }

 PrintMessage(Important,"Exiting.");

 FinishConfigurationFile();

#if !defined(__CYGWIN__)
 close(fSpoolDir);
#endif

 return(0);
}


/*++++++++++++++++++++++++++++++++++++++
  Print the program usage in long or short format.

  int verbose Set to 0 for short format, 1 for long format, 2 for version.
  ++++++++++++++++++++++++++++++++++++++*/

static void usage(int verbose)
{
 if(verbose==2)
   {
    fprintf(stderr,
            "wwwoffled version %s\n",
            WWWOFFLE_VERSION);
    exit(0);
   }

 fprintf(stderr,
         "\n"
         "WWWOFFLED - World Wide Web Offline Explorer (Daemon) - Version %s\n"
         "\n",WWWOFFLE_VERSION);

 if(verbose)
    fprintf(stderr,
            "(c) Andrew M. Bishop 1996,97,98,99,2000,01 [       amb@gedanken.demon.co.uk ]\n"
            "                                           [http://www.gedanken.demon.co.uk/]\n"
            "\n");

 fprintf(stderr,
         "Usage: wwwoffled -h | --help | --version\n"
         "       wwwoffled [-c <config-file>] [-d [<log-level>]]\n"
         "\n");

 if(verbose)
    fprintf(stderr,
            "   -h | --help         : Display this help.\n"
            "\n"
            "   --version           : Display the version number of wwwoffled.\n"
            "\n"
            "   -c <config-file>    : The name of the configuration file to use.\n"
            "                         (Default %s).\n"
            "\n"
            "   -d [<log-level>]    : Do not detach from the terminal and use stderr.\n"
            "                         0 <= log-level <= %d (default in config file).\n"
            "\n"
            "   -f                  : Do not fork children for each HTTP request.\n"
            "                         (Exits after first request, implies -d)\n"
            "\n"
            "   -p                  : Print the pid of wwwoffled on stdout (unless -d).\n"
            "\n",ConfigurationFileName(),Fatal+1);

 exit(0);
}


/*++++++++++++++++++++++++++++++++++++++
  Detach ourself from the controlling terminal.
  ++++++++++++++++++++++++++++++++++++++*/

static void demoninit(void)
{
 pid_t pid=0;

 pid=fork();
 if(pid==-1)
    PrintMessage(Fatal,"Cannot fork() to detach(1) [%!s].");
 else if(pid)
    exit(0);

 setsid();

 pid=fork();
 if(pid==-1)
    PrintMessage(Fatal,"Cannot fork() to detach(2) [%!s].");
 else if(pid)
   {
    /* print the child pid on stdout as requested */
    if(print_pid)
       printf("%d\n", pid);

    exit(0);
   }

 /* Close fds */
 close(STDIN_FILENO);
 close(STDOUT_FILENO);
 close(STDERR_FILENO);
}


/*++++++++++++++++++++++++++++++++++++++
  Install the signal handlers.
  ++++++++++++++++++++++++++++++++++++++*/

static void install_sighandlers(void)
{
 struct sigaction action;

 /* SIGCHLD */
 action.sa_handler = sigchild;
 sigemptyset (&action.sa_mask);
 action.sa_flags = 0;
 if(sigaction(SIGCHLD, &action, NULL) != 0)
    PrintMessage(Warning, "Cannot install SIGCHLD handler.");

 /* SIGINT, SIGQUIT, SIGTERM */
 action.sa_handler = sigexit;
 sigemptyset(&action.sa_mask);
 sigaddset(&action.sa_mask, SIGINT);           /* Block all of them */
 sigaddset(&action.sa_mask, SIGQUIT);
 sigaddset(&action.sa_mask, SIGTERM);
 action.sa_flags = 0;
 if(sigaction(SIGINT, &action, NULL) != 0)
    PrintMessage(Warning, "Cannot install SIGINT handler.");
 if(sigaction(SIGQUIT, &action, NULL) != 0)
    PrintMessage(Warning, "Cannot install SIGQUIT handler.");
 if(sigaction(SIGTERM, &action, NULL) != 0)
    PrintMessage(Warning, "Cannot install SIGTERM handler.");

 /* SIGHUP */
 action.sa_handler = sighup;
 sigemptyset(&action.sa_mask);
 action.sa_flags = 0;
 if(sigaction(SIGHUP, &action, NULL) != 0)
    PrintMessage(Warning, "Cannot install SIGHUP handler.");

 /* SIGPIPE */
 action.sa_handler = SIG_IGN;
 sigemptyset (&action.sa_mask);
 action.sa_flags = 0;
 if(sigaction(SIGPIPE, &action, NULL) != 0)
    PrintMessage(Warning, "Cannot ignore SIGPIPE.");
}


/*++++++++++++++++++++++++++++++++++++++
  The signal handler for the child exiting.

  int signum The signal number.
  ++++++++++++++++++++++++++++++++++++++*/

static void sigchild(/*@unused@*/ int signum)
{
 got_sigchld=1;
}


/*++++++++++++++++++++++++++++++++++++++
  The signal handler for the signals to tell us to exit.

  int signum The signal number.
  ++++++++++++++++++++++++++++++++++++++*/

static void sigexit(/*@unused@*/ int signum)
{
 closedown=1;
}


/*++++++++++++++++++++++++++++++++++++++
  The signal handler for the signals to tell us to re-read the config file.

  int signum The signal number.
  ++++++++++++++++++++++++++++++++++++++*/

static void sighup(/*@unused@*/ int signum)
{
 readconfig=1;
}
